### PR TITLE
Use BUILD_WORKSPACE_DIRECTORY in SwiftLintPlugin

### PR DESCRIPTION
**Pull request by garricn**
## Changelog

- Rewrite `SwiftLintPlugin` using `BUILD_WORKSPACE_DIRECTORY` without relying on the `--config` option
- Rewrite `Plugins/SwiftLintPlugin/Path+Helpers.swift`
- Update Plug-in Support in `README.md`

## Benefits

- Supports parent/child configurations
- Supports nested configurations

## Background

The current version of the plugin uses the `--config` option which is incompatible with project structures using parent/child configurations or nested configurations.

- References:
    - https://github.com/realm/SwiftLint/blob/0.50.3/Plugins/SwiftLintPlugin/SwiftLintPlugin.swift#L33
    - https://github.com/realm/SwiftLint/blob/0.50.3/Plugins/SwiftLintPlugin/SwiftLintPlugin.swift#L79

Instead of relying on the `--config` option, the plugin can leverage the `BUILD_WORKSPACE_DIRECTORY` environment variable (recently added to support Bazel) which sets SwiftLint’s working directory.

- Reference: https://github.com/realm/SwiftLint/blob/0.50.3/Source/swiftlint/Commands/SwiftLint.swift#L7

## Implementation Details

The plugin determines the SwiftLint working directory by locating the topmost config file within the package/project directory. If a config file is not found therein, the package/project directory is used as the working directory.

## Project Structures

The plugin throws an error when it is unable to resolve the SwiftLint working directory. For example, this will occur in Xcode projects where the target's Swift files are not located within the project directory.

To maximize compatibility with the plugin, avoid project structures that require the use of the `--config` option.

### Unexpected Project Structures

Project structures where SwiftLint's configuration file is located outside of the package/project directory are not directly supported by the build tool plugin. This is because it isn't possible to pass arguments to build tool plugins (e.g., passing the config file path). 

If your project structure doesn't work directly with the build tool plugin, please consider one of the following options:

- To use a config file located outside the package/project directory, a config file may be added to that directory specifying a parent config path to the other config file, e.g., `parent_config: path/to/.swiftlint.yml`.
- You can also consider the use of a Run Script Build Phase in place of the build tool plugin.

### CODE REVIEW

**Comment by tonyarnold** [comment]
To clarify `BUILD_WORKSPACE_DIRECTORY` is not available within Xcode, or within standard Apple toolchains - what happens to those configurations?

**Comment by tonyarnold** [comment]
I'm happy to see a cleaner option for configuration discovery! 

Also happy to see #4757 rolled into this.

**Comment by tonyarnold** [comment]
I've just given this a run, and it's not working for Xcode targets: 

```
working directory:
   /Users/tonyarnold/Developer/Itty Bitty Apps/Reveal/Application
tool mapping:
   swiftlint: /Users/tonyarnold/Library/Developer/Xcode/DerivedData/Reveal-ezgnpwiedoerqlgbbonoifrichyk/SourcePackages/artifacts/swiftlint/SwiftLintBinary.artifactbundle/swiftlint-0.50.3-macos/bin/swiftlint
tool search paths:
   /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
   /Applications/Xcode.app/Contents/Developer/usr/bin
   /bin
   /sbin
   /usr/bin
   /usr/sbin

failedToResolveSwiftLintWorkingDirectory
```

In this instance, the `.swiftlint.yml` is held in `/Users/tonyarnold/Developer/Itty Bitty Apps/Reveal`, which is the parent of the Xcode project directory.

**Comment by garricn** [comment]
@tonyarnold Thank you for your comments and for giving it a try! The Plugin does work for Xcode targets, but only when the config file is located within "the package/project directory (and its sub-directories)" as mentioned in the PR description.

We took this approach as it ensures support for the most significant number of predictable project structures.

We can consider traversing the filesystem above the project directory as you suggest. Before we do that, however, we'd like to propose using symlinks for unpredictable project structures, such as when the config file is contained in an ancestor or sibling directory. 

We've successfully tested this approach with an example project matching your setup.

To test this, a symlink that links to the config file in the parent directory should be added to each of your project directories, i.e. `../.swiftlint.yml`.

If your test is successful, we will update the `Project Structures` section of the PR description with the following:

```
To provide the plugin access to a config file located outside the package/project directory, 
a symlink that links to the config file should be added to the package/project directory.
```

We're excited to receive the results of your test. Thank you!

**Comment by garricn** [comment]
@tonyarnold We've gone ahead and proactively updated the Plug-in Support section of the README.md including the information about the symlink solution. We've updated the PR description to match.

**Comment by tonyarnold** [comment]
Thanks @garricn - I've just tried using this branch, with symbolic links to the "root" `.swiftlint.yml` in the root of my repository/workspace, but sadly the workaround you've suggested doesn't work for me - the plugin still throws `failedToResolveSwiftLintWorkingDirectory`.

**Comment by tonyarnold** [comment]
I couldn't get the symlinks to work reliably, so I placed a `.swiftlint.yml` file in each of my project directories containing:

```
parent_config: ../.swiftlint.yml
```

I think it's messy, but it solves the problem.

**Comment by tonyarnold** [comment]
Does running the SwiftLint CLI in a directory without a configuration file throw an error? It'd be a pretty rare combination to just accept the out-of-the-box defaults, right?

**Comment by garricn** [comment]
> I couldn't get the symlinks to work reliably, so I placed a `.swiftlint.yml` file in each of my project directories containing:
> 
> ```
> parent_config: ../.swiftlint.yml
> ```
> 
> I think it's messy, but it solves the problem.

Hey @tonyarnold, we appreciate that and understand that the parent config option can seem messy for a setup like yours. However, we believe that it's an ideal approach for the plugin since it's a supported feature of SwiftLint. Thank you for continuing to test this out and for providing feedback. We've updated the README and the PR description with information about using parent configs when needed.

**Comment by garricn** [comment]
> Does running the SwiftLint CLI in a directory without a configuration file throw an error? It'd be a pretty rare combination to just accept the out-of-the-box defaults, right?

Hey @tonyarnold, we tested this locally and a config file is not required. We updated the plugin to run SwiftLint in the package/project directory if it's unable to locate a working directory containing a config file. Thank you for pointing this out!

**Comment by tonyarnold** [comment]
Awesome, this is looking great and it will work for what I use the plugin for based on my testing. Thanks for the work you've put into this!

Do you have a team at your end working on this, @garricn? Gosh, if only all OSS contributors were so lucky!

**Comment by garricn** [comment]
> Awesome, this is looking great and it will work for what I use the plugin for based on my testing. Thanks for the work you've put into this!
> 
> Do you have a team at your end working on this, @garricn? Gosh, if only all OSS contributors were so lucky!

Thank you for your recognition @tonyarnold. We really appreciate it and are glad this solution will work for you.

We have not had many opportunities to contribute to open-source projects. However, our team is working on internal Swift packages and SwiftLint is crucial for maintaining quality and consistency across our Swift repositories. As our internal contributor base has grown, we identified that it is essential to have violations appear in Xcode. Once we discovered that parent/child and nested configurations (which we rely on) were not being respected, we immediately committed ourselves to improve the plugin.

I assume you're asking about the team because I keep using "we". I'm working on this @chrisfuller.

TL;DR … this contribution is of particular importance to our team and the projects we work on and we hope it gets merged soon. 🤞🏼

**Review by tonyarnold — APPROVED** [review]
The code is great, and well documented. 

I've made a few suggestions about the wording of the documentation. "Unpredictable" isn't the word I would use to describe the project structures that don't work out of the box here - they're unexpected, perhaps? 

Thanks for the hard work on this! Looking forward to seeing it merged,

**Comment by garricn** [comment]
> The code is great, and well documented.
> 
> I've made a few suggestions about the wording of the documentation. "Unpredictable" isn't the word I would use to describe the project structures that don't work out of the box here - they're unexpected, perhaps?
> 
> Thanks for the hard work on this! Looking forward to seeing it merged 👍

Thank you for your contributions @tonyarnold. We're also looking forward to getting it merged.

**Comment by tinder-garricnahapetian** [comment]
Hello @jpsim @marcelofabri @SimplyDanny 👋 

It's been a few months so I wanted to check in again. What are the next steps for this PR? Please let me know as we are eager to get it merged in. Thank you 😄

**Comment by garricn** [comment]
Hi @SimplyDanny. Thank you for approving and merging my other two recent PRs. May you please review and merge this PR as well 🙏 Let me know what you think. Thank you 😄

File: README.md [hunk 1]
```diff
@@ -200,12 +200,36 @@ If you've installed SwiftLint via CocoaPods the script should look like this:
...
+the topmost config file within the package/project directory. If a config file
+is not found therein, the package/project directory is used as the working 
+directory.
+
+The plugin throws an error when it is unable to resolve the SwiftLint working 
+directory. For example, this will occur in Xcode projects where the target's
+Swift files are not located within the project directory.
+
+To maximize compatibility with the plugin, try to avoid project structures that 
+require the use of the `--config` option.
```
Comment (garricn) [inline README.md:216]:
```suggestion
To maximize compatibility with the plugin, avoid project structures that require
the use of the `--config` option.
```

**Comment by tinder-cfuller** [comment]
@garricn This change is very much welcomed and appreciated. Thank you for submitting this!

I see that it has been open for 1 year. I would like to kindy mention the project maintainers, @jpsim @SimplyDanny and @marcelofabri, to please comment and provide guidance. I find this change to be ideal and would like to see this merged in. But ask the maintainers to please clarify why this change has not been been accepted yet. Does it need further changes? Or should the PR simply be closed at this point? Asking since knowing that it will not be merged is very helpful as well – so then I can focus on creating an alternative custom solution for the packages I maintain.

Thank you!

CC: @tonyarnold

File: Plugins/SwiftLintPlugin/SwiftLintPlugin.swift [hunk 2]
```diff
@@ -3,81 +3,146 @@ import PackagePlugin
...
-
         // Determine whether we need to enable cache or not (for Xcode Cloud we don't)
+        let cacheArguments: [String]
         if ProcessInfo.processInfo.environment["CI_XCODE_CLOUD"] == "TRUE" {
-            arguments.append("--no-cache")
+            cacheArguments = ["--no-cache"]
         } else {
-            arguments.append("--cache-path")
-            arguments.append("\(workingDirectory)")
+            let cachePath: Path = path.appending("Cache")
```
Comment (SimplyDanny) [inline Plugins/SwiftLintPlugin/SwiftLintPlugin.swift:66]: Should this be a hidden directory like `.cache`.

File: Plugins/SwiftLintPlugin/SwiftLintPlugin.swift [hunk 3]
```diff
@@ -3,81 +3,146 @@ import PackagePlugin
...
+            cacheArguments = ["--cache-path", "\(cachePath)"]
         }
-
-        // Manually look for configuration files, to avoid issues when the plugin does not execute our tool from the
-        // package source directory.
-        if let configuration = packageDirectory.firstConfigurationFileInParentDirectories() {
-            arguments.append(contentsOf: ["--config", "\(configuration.string)"])
+        let outputPath: Path = path.appending("Output")
+        do {
+            try FileManager.default.createDirectory(atPath: outputPath.string, withIntermediateDirectories: true)
```
Comment (SimplyDanny) [inline Plugins/SwiftLintPlugin/SwiftLintPlugin.swift:76]: What is this output folder going to be used for?

File: Plugins/SwiftLintPlugin/SwiftLintPlugin.swift [hunk 4]
```diff
@@ -3,81 +3,146 @@ import PackagePlugin
...
+        } catch {
+            print(error.localizedDescription)
```
Comment (SimplyDanny) [inline Plugins/SwiftLintPlugin/SwiftLintPlugin.swift:78]: Is printing the error sufficient? What if the `Output` folder doesn't exist while running the plugin?

File: Plugins/SwiftLintPlugin/SwiftLintPluginError.swift [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    case swiftFilesNotInProjectDirectory(Path)
+    case swiftFilesNotInWorkingDirectory(Path)
+
+    var description: String {
+        switch self {
+        case let .pathNotInDirectory(path, directory):
+            return """
+                Path is not in directory.
+                Path: \(path)
+                Directory: \(directory)
```
Comment (SimplyDanny) [inline Plugins/SwiftLintPlugin/SwiftLintPluginError.swift:14]:
```suggestion
                Path '\(path)' is not in directory '\(directory)'.
```

File: Plugins/SwiftLintPlugin/SwiftLintPluginError.swift [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+                """
+        case let .swiftFilesNotInProjectDirectory(directory):
+            return """
+                Swift files are not in project directory.
+                Directory: \(directory)
```
Comment (SimplyDanny) [inline Plugins/SwiftLintPlugin/SwiftLintPluginError.swift:19]:
```suggestion
                Swift files are not in project directory '\(directory)'.
```

File: Plugins/SwiftLintPlugin/SwiftLintPluginError.swift [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+                """
+        case let .swiftFilesNotInWorkingDirectory(directory):
+            return """
+                Swift files are not in working directory.
+                Directory: \(directory)
```
Comment (SimplyDanny) [inline Plugins/SwiftLintPlugin/SwiftLintPluginError.swift:24]:
```suggestion
                Swift files are not in working directory '\(directory)'.
```

**Review by SimplyDanny — COMMENTED** [review]
First of all: Sorry for the very late response! The problem is that none of the maintainers actually uses the plugin. So we cannot really judge on these kinds of changes convincingly.

I'm fine with merging it once a few comprehensive questions are clarified and some nits are fixed.

It would be really great if you guys can jump in if any issues with the plugin come up in the future. With your experience you are more capable of answering question, provide some help or propose fixes.

File: Plugins/SwiftLintPlugin/SwiftLintPlugin.swift (hunk 4 above)
Comment (garricn) [reply Plugins/SwiftLintPlugin/SwiftLintPlugin.swift:78]: Great catch. Will update to throw instead.

File: Plugins/SwiftLintPlugin/SwiftLintPlugin.swift (hunk 3 above)
Comment (garricn) [reply Plugins/SwiftLintPlugin/SwiftLintPlugin.swift:76]: The output directory, although not used by SwiftLint, is a required argument when creating the `.prebuildCommand(...)`.

File: Plugins/SwiftLintPlugin/SwiftLintPlugin.swift (hunk 2 above)
Comment (garricn) [reply Plugins/SwiftLintPlugin/SwiftLintPlugin.swift:66]: The `PluginContext` provides the work directory (`path`) to the plugin to use as necessary. Two subdirectories (one named `Cache` and the other named `Output`) are created. The directories do not need to be hidden since they are within the work directory. The cache directory is used for SwiftLint's cache. The output directory, although not used by SwiftLint, is a required argument when creating the `.prebuildCommand(...)`.

**Comment by tinder-garricnahapetian** [comment]
> First of all: Sorry for the very late response! Thank you for your patience and continuous reminders. The problem is that none of the maintainers actually uses the plugin. So we cannot really judge on these kinds of changes convincingly.
> 
> I'm fine with merging it once a few comprehensive questions are clarified and some nits are fixed.
> 
> It would be really great if you guys can jump in if any issues with the plugin come up in the future. With your experience you are more capable of answering question, provide some help or propose fixes.

@SimplyDanny Thank you so much! This is really great to hear. Please feel free to mention me on any future issues that arise with the plugin.

File: Plugins/SwiftLintPlugin/SwiftLintPluginError.swift [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+    var description: String {
+        switch self {
+        case let .pathNotInDirectory(path, directory):
+            return "Path '\(path)' is not in directory '\(directory)'."
+        case let .swiftFilesNotInProjectDirectory(directory):
+            return "Swift files are not in project directory '\(directory)'."
+        case let .swiftFilesNotInWorkingDirectory(directory):
+            return "Swift files are not in working directory '\(directory)'."
+        }
```
Comment (SimplyDanny) [inline Plugins/SwiftLintPlugin/SwiftLintPluginError.swift:16]: You may use a `switch` expression:
```suggestion
        switch self {
        case let .pathNotInDirectory(path, directory):
            "Path '\(path)' is not in directory '\(directory)'."
        case let .swiftFilesNotInProjectDirectory(directory):
            "Swift files are not in project directory '\(directory)'."
        case let .swiftFilesNotInWorkingDirectory(directory):
            "Swift files are not in working directory '\(directory)'."
        }
```

File: Plugins/SwiftLintPlugin/SwiftLintPlugin.swift [hunk 9]
```diff
@@ -3,81 +3,138 @@ import PackagePlugin
...
+    private func environment(
+        context: XcodePluginContext,
+        target: XcodeTarget
+    ) throws -> [String: String] {
+        let projectDirectory: Path = context.xcodeProject.directory
+        let swiftFiles: [Path] = swiftFiles(target: target)
+        let swiftFilesNotInProjectDirectory: [Path] = swiftFiles.filter { !$0.isDescendant(of: projectDirectory) }
+
+        guard swiftFilesNotInProjectDirectory.isEmpty
+        else { throw SwiftLintPluginError.swiftFilesNotInProjectDirectory(projectDirectory) }
```
Comment (SimplyDanny) [inline Plugins/SwiftLintPlugin/SwiftLintPlugin.swift:127]: This style is uncommon for this project.
```suggestion
        guard swiftFilesNotInProjectDirectory.isEmpty else { 
            throw SwiftLintPluginError.swiftFilesNotInProjectDirectory(projectDirectory)
        }
```

File: CHANGELOG.md [hunk 10]
```diff
@@ -838,6 +838,11 @@
 * Catch more valid `no_magic_numbers` violations.  
   [JP Simard](https://github.com/jpsim)
 
+* Rewrite `SwiftLintPlugin` using `BUILD_WORKSPACE_DIRECTORY` without relying
+  on the `--config` option.  
+  [Garric Nahapetian](https://github.com/garricn)
+  [#4758](https://github.com/realm/SwiftLint/pull/4758)
```
Comment (SimplyDanny) [inline CHANGELOG.md:844]: Omit this line when there's no associated issue:
```suggestion
```
The PR number will be attached to the commit message anyway.

**Review by SimplyDanny — APPROVED** [review]

**Comment by tinder-cfuller** [comment]
Thank you @SimplyDanny! ❤️
